blobstore/upload.go

```diff
@@ -220,33 +207,223 @@ func (s3Ctrl *S3Controller) GetUploadPresignedURL(bucket string, key string, exp
 	return urlStr, nil
 }
 
+// function to retrieve presigned url for a multipart upload part.
+func (s3Ctrl *S3Controller) GetUploadPartPresignedURL(bucket string, key string, uploadID string, partNumber int64, expMin int) (string, error) {
+	duration := time.Duration(expMin) * time.Minute
+	req, _ := s3Ctrl.S3Svc.UploadPartRequest(&s3.UploadPartInput{
+		Bucket:     aws.String(bucket),
+		Key:        aws.String(key),
+		UploadId:   aws.String(uploadID),
+		PartNumber: aws.Int64(partNumber),
+	})
+
+	urlStr, err := req.Presign(duration)
+	if err != nil {
+		return "", err
+	}
+
+	return urlStr, nil
+}
+
+// enpoint handler that will either return a one time presigned upload URL or multipart upload url
 func (bh *BlobHandler) HandleGetPresignedUploadURL(c echo.Context) error {
+	key := c.QueryParam("key")
+	if key == "" {
+		errMsg := fmt.Errorf("`key` parameters are required")
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+	bucket := c.QueryParam("bucket")
```

Comment:
Also check if bucket == "" ? or move bh.GetController before checking the write permission, otherwise it will say a user does not have permission because of a bad request

Reply:
I refactored that check in `bh.GetController,  since I was doing that a lot, I also thought that there wouldn't be a point in returning a controller if the security check fails. I'll move the controller-getter before the security check.